File: android/app/src/main/java/com/openburnbar/data/policy/MobileStoreEntitlementPolicy.kt
package com.openburnbar.data.policy

enum class MobileStoreProductRole {
    CLOUD_MONTHLY,
    CLOUD_ANNUAL,
    PRO_MONTHLY,
    PRO_ANNUAL,
    ULTRA_MONTHLY,
    ULTRA_ANNUAL,
    AGENT_CONTROL_TOP_UP,
    FLOO_RELAY_TOP_UP,
    ELDER_WAND_100,
    ELDER_WAND_500,
}

enum class MobileStoreEntitlementState(val wire: String) {
    NONE("none"),
    ACTIVE("active"),
    EXPIRED("expired"),
    REFUNDED("refunded"),
    REVOKED("revoked"),
    MISSING_CATALOG("missing-catalog"),
    RESTORE_PENDING("restore-pending"),
}

sealed class MobileStorePriceDisplay {
    data class Live(val price: String) : MobileStorePriceDisplay()
    data object Unavailable : MobileStorePriceDisplay()

    val customerFacingText: String
        get() = when (this) {
            is Live -> price
            Unavailable -> MobileStoreEntitlementPolicy.UNAVAILABLE_PRICE_LABEL
        }

    val isLivePrice: Boolean get() = this is Live
}

object MobileStoreEntitlementPolicy {
    const val UNAVAILABLE_PRICE_LABEL = "Price unavailable"

    const val APPLE_CLOUD_MONTHLY = "com.openburnbar.pro.monthly"
    const val APPLE_CLOUD_ANNUAL = "com.openburnbar.pro.annual"
    const val APPLE_PRO_MONTHLY = "com.openburnbar.proMax.v2.monthly"
    const val APPLE_PRO_ANNUAL = "com.openburnbar.proMax.annual"
    const val APPLE_ULTRA_MONTHLY = "com.openburnbar.ultra.monthly"
    const val APPLE_ULTRA_ANNUAL = "com.openburnbar.ultra.annual.v2"
    const val APPLE_AGENT_CONTROL = "com.openburnbar.agentControl.actions100"
    const val APPLE_FLOO_RELAY = "com.openburnbar.floo.relay50gb"
    const val APPLE_ELDER_WAND_100 = "com.openburnbar.elderWand.searches100"
    const val APPLE_ELDER_WAND_500 = "com.openburnbar.elderWand.searches500"

    const val PLAY_CLOUD_MONTHLY = "com.openburnbar.pro.monthly"
    const val PLAY_CLOUD_ANNUAL = "com.openburnbar.pro.annual"
    const val PLAY_PRO_MONTHLY = "com.openburnbar.promax.v2.monthly"
    const val PLAY_PRO_ANNUAL = "com.openburnbar.promax.annual"
    const val PLAY_ULTRA_MONTHLY = "com.openburnbar.ultra.monthly"
    const val PLAY_ULTRA_ANNUAL = "com.openburnbar.ultra.annual"
    const val PLAY_AGENT_CONTROL = "com.openburnbar.agentcontrol.actions100"
    const val PLAY_FLOO_RELAY = "com.openburnbar.floo.relay50gb"
    const val PLAY_ELDER_WAND_100 = "com.openburnbar.elderwand.searches100"
    const val PLAY_ELDER_WAND_500 = "com.openburnbar.elderwand.searches500"

    fun role(productID: String): MobileStoreProductRole? = when (productID) {
        APPLE_CLOUD_MONTHLY, PLAY_CLOUD_MONTHLY -> MobileStoreProductRole.CLOUD_MONTHLY
        APPLE_CLOUD_ANNUAL, PLAY_CLOUD_ANNUAL -> MobileStoreProductRole.CLOUD_ANNUAL
        APPLE_PRO_MONTHLY, PLAY_PRO_MONTHLY -> MobileStoreProductRole.PRO_MONTHLY
        APPLE_PRO_ANNUAL, PLAY_PRO_ANNUAL -> MobileStoreProductRole.PRO_ANNUAL
        APPLE_ULTRA_MONTHLY, PLAY_ULTRA_MONTHLY -> MobileStoreProductRole.ULTRA_MONTHLY
        APPLE_ULTRA_ANNUAL, PLAY_ULTRA_ANNUAL -> MobileStoreProductRole.ULTRA_ANNUAL
        APPLE_AGENT_CONTROL, PLAY_AGENT_CONTROL -> MobileStoreProductRole.AGENT_CONTROL_TOP_UP
        APPLE_FLOO_RELAY, PLAY_FLOO_RELAY -> MobileStoreProductRole.FLOO_RELAY_TOP_UP
        APPLE_ELDER_WAND_100, PLAY_ELDER_WAND_100 -> MobileStoreProductRole.ELDER_WAND_100
        APPLE_ELDER_WAND_500, PLAY_ELDER_WAND_500 -> MobileStoreProductRole.ELDER_WAND_500
        else -> null
    }

    fun displayPrice(livePrice: String?): MobileStorePriceDisplay {
        val trimmed = livePrice?.trim().orEmpty()
        return if (trimmed.isEmpty() || trimmed == UNAVAILABLE_PRICE_LABEL) {
            MobileStorePriceDisplay.Unavailable
        } else {
            MobileStorePriceDisplay.Live(trimmed)
        }
    }

    fun classify(
        catalogPresent: Boolean,
        restoring: Boolean,
        revoked: Boolean,
        refunded: Boolean,
        expired: Boolean,
        active: Boolean,
    ): MobileStoreEntitlementState = when {
        !catalogPresent -> MobileStoreEntitlementState.MISSING_CATALOG
        restoring -> MobileStoreEntitlementState.RESTORE_PENDING
        revoked -> MobileStoreEntitlementState.REVOKED
        refunded -> MobileStoreEntitlementState.REFUNDED
        expired -> MobileStoreEntitlementState.EXPIRED
        active -> MobileStoreEntitlementState.ACTIVE
        else -> MobileStoreEntitlementState.NONE
    }
}

File: OpenBurnBarCore/Sources/OpenBurnBarMobilePolicy/MobileStoreEntitlementPolicy.swift
import Foundation

/// StoreKit / Play product IDs that must stay semantically aligned.
public enum MobileStoreProductRole: String, Sendable, Equatable, CaseIterable {
    case cloudMonthly
    case cloudAnnual
    case proMonthly
    case proAnnual
    case ultraMonthly
    case ultraAnnual
    case agentControlTopUp
    case flooRelayTopUp
    case elderWand100
    case elderWand500
}

public enum MobileStoreEntitlementState: String, Sendable, Equatable, CaseIterable {
    case none
    case active
    case expired
    case refunded
    case revoked
    case missingCatalog = "missing-catalog"
    case restorePending = "restore-pending"
}

public enum MobileStorePriceDisplay: Sendable, Equatable {
    case live(String)
    case unavailable

    public var customerFacingText: String {
        switch self {
        case .live(let price): return price
        case .unavailable: return MobileStoreEntitlementPolicy.unavailablePriceLabel
        }
    }

    public var isLivePrice: Bool {
        if case .live = self { return true }
        return false
    }
}

public enum MobileStoreEntitlementPolicy {
    /// Not a price. Shown only when StoreKit / Play has not returned ProductDetails.
    public static let unavailablePriceLabel = "Price unavailable"

    public static let appleCloudMonthly = "com.openburnbar.pro.monthly"
    public static let appleCloudAnnual = "com.openburnbar.pro.annual"
    public static let appleProMonthly = "com.openburnbar.proMax.v2.monthly"
    public static let appleProAnnual = "com.openburnbar.proMax.annual"
    public static let appleUltraMonthly = "com.openburnbar.ultra.monthly"
    public static let appleUltraAnnual = "com.openburnbar.ultra.annual.v2"
    public static let appleAgentControl = "com.openburnbar.agentControl.actions100"
    public static let appleFlooRelay = "com.openburnbar.floo.relay50gb"
    public static let appleElderWand100 = "com.openburnbar.elderWand.searches100"
    public static let appleElderWand500 = "com.openburnbar.elderWand.searches500"

    public static let playCloudMonthly = "com.openburnbar.pro.monthly"
    public static let playCloudAnnual = "com.openburnbar.pro.annual"
    public static let playProMonthly = "com.openburnbar.promax.v2.monthly"
    public static let playProAnnual = "com.openburnbar.promax.annual"
    public static let playUltraMonthly = "com.openburnbar.ultra.monthly"
    public static let playUltraAnnual = "com.openburnbar.ultra.annual"
    public static let playAgentControl = "com.openburnbar.agentcontrol.actions100"
    public static let playFlooRelay = "com.openburnbar.floo.relay50gb"
    public static let playElderWand100 = "com.openburnbar.elderwand.searches100"
    public static let playElderWand500 = "com.openburnbar.elderwand.searches500"

    public static func role(for productID: String) -> MobileStoreProductRole? {
        switch productID {
        case appleCloudMonthly, playCloudMonthly: return .cloudMonthly
        case appleCloudAnnual, playCloudAnnual: return .cloudAnnual
        case appleProMonthly, playProMonthly: return .proMonthly
        case appleProAnnual, playProAnnual: return .proAnnual
        case appleUltraMonthly, playUltraMonthly: return .ultraMonthly
        case appleUltraAnnual, playUltraAnnual: return .ultraAnnual
        case appleAgentControl, playAgentControl: return .agentControlTopUp
        case appleFlooRelay, playFlooRelay: return .flooRelayTopUp
        case appleElderWand100, playElderWand100: return .elderWand100
        case appleElderWand500, playElderWand500: return .elderWand500
        default: return nil
        }
    }

    public static func displayPrice(livePrice: String?) -> MobileStorePriceDisplay {
        let trimmed = livePrice?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        guard !trimmed.isEmpty, trimmed != unavailablePriceLabel else {
            return .unavailable
        }
        return .live(trimmed)
    }

    public static func classify(
        catalogPresent: Bool,
        restoring: Bool,
        revoked: Bool,
        refunded: Bool,
        expired: Bool,
        active: Bool
    ) -> MobileStoreEntitlementState {
        if !catalogPresent { return .missingCatalog }
        if restoring { return .restorePending }
        if revoked { return .revoked }
        if refunded { return .refunded }
        if expired { return .expired }
        if active { return .active }
        return .none
    }
}
